### DISCUSSION TITLE
Custom Headers for 404 pages

### DISCUSSION THREAD

**Discussion by Antonio-Laguna** [vercel/next.js#70624] [App Router]
Is there a way to set custom headers for 404 errors? Use case is to lower the cache on the CDN so if it eventually becomes visible it works. 

Right now only solution that comes to mind is to create a custom server but feels too much

**Comment by MkMonir** [comment]
You can try this code - 

// next.config.js
module.exports = {
  async headers() {
    return [
      {
        source: '/404', // target your 404 page
        headers: [
          {
            key: 'Cache-Control',
            value: 'no-cache, no-store, must-revalidate', // Set low/no caching for 404
          },
          {
            key: 'Pragma',
            value: 'no-cache',
          },
          {
            key: 'Expires',
            value: '0',
          },
        ],
      },
    ];
  },
};

// pages/404.js
export async function getServerSideProps({ res }) {
  res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
  return { props: {} };
}

export default function Custom404() {
  return <h1>404 - Page Not Found</h1>;
}

**Comment by Antonio-Laguna** [comment]
@MkMonir that could work if it wasn't because 404 doesn't really get you to any given page and stay in place

**Comment by MkMonir** [comment]
@Antonio-Laguna Then you can use middleware for that - 

import { NextResponse } from 'next/server';

export function middleware(request) {
  const response = NextResponse.next();

  // If the response status is 404, set custom cache headers
  if (request.nextUrl.pathname !== '/404' && !request.nextUrl.pathname.startsWith('/api')) {
    response.headers.set('Cache-Control', 'no-cache, no-store, must-revalidate');
    response.headers.set('Pragma', 'no-cache');
    response.headers.set('Expires', '0');
  }

  return response;
}

export const config = {
  matcher: '/:path*', // Apply to all routes
};

**Comment by Antonio-Laguna** [comment]
How's that checking the response was a 404?
